services/gitea/gitea-base.js

```diff
@@ -0,0 +1,90 @@
+import { BaseJsonService } from '../index.js'
+
+export default class GiteaBase extends BaseJsonService {
+  static auth = {
+    passKey: 'gitea_token',
+    serviceKey: 'gitea',
+  }
+
+  async fetch({ url, options, schema, httpErrors }) {
+    return this._requestJson(
+      this.authHelper.withBearerAuthHeader({
+        schema,
+        url,
+        options,
+        httpErrors,
+      }),
+    )
+  }
+
+  async fetchPage({ page, requestParams, schema }) {
+    const {
+      options: { searchParams: existingQuery, ...restOptions } = {},
+      ...rest
+    } = requestParams
+
+    requestParams = {
+      options: {
+        searchParams: {
+          ...existingQuery,
+          ...{ page },
+        },
+        ...restOptions,
+      },
+      ...rest,
+    }
+
+    const { res, buffer } = await this._request(
+      this.authHelper.withBearerAuthHeader({
+        ...requestParams,
+      }),
+    )
+
+    const json = this._parseJson(buffer)
+    const data = this.constructor._validate(json, schema)
+
+    return { res, data }
+  }
+
+  async fetchPaginatedArrayData({
+    url,
+    options,
+    schema,
+    httpErrors,
+    firstPageOnly = false,
+  }) {
+    const requestLimit = 100
+    const requestParams = {
+      url,
+      options: {
+        headers: { Accept: 'application/json' },
+        searchParams: { limit: requestLimit },
+        ...options,
+      },
+      httpErrors,
+    }
+
+    const {
+      res: { headers },
+      data,
+    } = await this.fetchPage({ page: 1, requestParams, schema })
+    const numberOfItems = headers['x-total-count']
+
+    let numberOfPages = 1
+    if (numberOfItems > 0) {
+      numberOfPages = Math.ceil(numberOfItems / requestLimit)
```

Comment:
Hmm. I have not had a really detailed look at this pagination code, but I'm not sure this is going to work.
If we look at
https://docs.gitea.com/next/development/api-usage
(which I think is relevant)

> `MAX_RESPONSE_ITEMS` is set to 50

says to me that

a) The default max allowable value of `requestLimit` for a gitea server is 50
b) This is a number that is going to be different for different gitea servers, so we can't safely assume its value

so I think trying to assume we can work out the number of pages up-front based on our hard-coded value of `requestLimit` and the total number of items feels like it is logic based on a bad assumption.

Tbh, the way we do this for the GitHub badges is we just request the first page of releases ordered by date, and if you want the latest semver then the 'search space' for that is the first page of results, mostly for performance reasons. We've yet to have anyone pop up whose latest release by semver isn't in the first page of results.

Tbh, I'd be happy with that solution here too. When I flagged that `fetchPage` was unused code, I assumed we'd probably just delete it.

Given we have to deal with multiple different implementations, if we do want to try and implement pagination like this, we'd need to infer the page size rather than assuming it up-front, but as I say, I'd also be happy to just gloss over it if we can.

Reply:
We have MAX_RESPONSE_ITEMS for the absolute maximum, and DEFAULT_PAGING_NUM for the default pagination value which can be different per every instance. We do get a header for `x-total-count` for how many items are available, and if this is greater than the number of items in the array we could calculate the pages.

When I found the pagination was not working in Gitlab's code too I thought to myself... who even has more than 100 releases to even scan through to find the latest SemVer! Happy to put this back and just assume the first page too.